packages/fxa-auth-server/lib/routes/linked-accounts.ts

```diff
@@ -37,64 +37,112 @@ export class LinkedAccountHandler {
+      case 'google': {
+        if (!this.googleAuthClient) {
+          throw error.thirdPartyAccountError();
+        }
+
+        const { clientId, clientSecret, redirectUri } =
+          this.config.googleAuthConfig;
+        let rawIdToken;
+        if (code) {
+          const data = {
+            code,
+            client_id: clientId,
+            client_secret: clientSecret,
+            redirect_uri: redirectUri,
+            grant_type: 'authorization_code',
+          };
+
+          try {
+            const res = await axios.post(
+              this.config.googleAuthConfig.tokenEndpoint,
+              data
+            );
+            // We currently only use the `id_token` after completing the
+            // authorization code exchange. In the future we could store a
+            // refresh token to do other things like revoking sessions.
+            //
+            // See https://developers.google.com/identity/protocols/oauth2/openid-connect#exchangecode
+            rawIdToken = res.data['id_token'];
+
+            const verifiedToken = await this.googleAuthClient.verifyIdToken({
+              idToken: rawIdToken,
+              audience: clientId,
+            });
+
+            idToken = verifiedToken.getPayload();
+          } catch (err) {
+            this.log.error('linked_account.code_exchange_error', err);
+            throw error.thirdPartyAccountError();
+          }
+        }
+        break;
       }
-    }
+      case 'apple': {
+        const { clientId, clientSecret } = this.config.appleAuthConfig;
+        const code = requestPayload.code;
+        if (code) {
+          const data = {
+            code,
+            client_id: clientId,
+            client_secret: clientSecret,
+            grant_type: 'authorization_code',
+          };
 
-    const verifiedToken = await this.googleAuthClient.verifyIdToken({
-      idToken,
-      audience: clientId,
-    });
+          try {
+            const res = await axios.post(
+              this.config.appleAuthConfig.tokenEndpoint,
+              new URLSearchParams(data).toString()
+            );
+            idToken = parseJwt(res.data['id_token']);
+          } catch (err) {
+            this.log.error('linked_account.code_exchange_error', err);
+            throw error.thirdPartyAccountError();
+          }
+        }
+        break;
+      }
+    }
 
-    const payload = verifiedToken.getPayload();
-    if (!payload) {
+    if (!idToken) {
       throw error.thirdPartyAccountError();
     }
-    const userid = payload['sub'];
+
+    const userid = idToken.sub;
+    const email = idToken.email;
 
     let accountRecord;
-    let googleRecord = await this.db.getGoogleId(userid);
+    let linkedAccountRecord = await this.db.getLinkedAccount(userid, provider);
 
-    if (!googleRecord) {
-      const email = payload.email;
+    if (!linkedAccountRecord) {
       try {
-        // This is a new Google account linking an existing FxA account
```

Comment:
It might be helpful to leave something similar to this comment like, "This is a new third party account linking an existing FxA account"